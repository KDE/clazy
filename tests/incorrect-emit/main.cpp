#include <QtCore/QObject>

class MyObject;

MyObject* somefunc()
{
    return nullptr;
}

static MyObject * s_obj;

class MyObject : public QObject
{
public:
    MyObject();
    void pub();
    MyObject* memberFunc() const;
    MyObject *another;
private:
    void priv();

public slots:
    void pubSlot();

signals:
    void sig();

private Q_SLOTS:
    void privSlot();

protected:
    void prot();
    Q_SIGNAL void singularSig();
    Q_SLOT void singularSlot();
};

void MyObject::pub()
{
    emit  prot(); // Warning: emit on non slot.
    sig(); // Warning: missing emit
    prot(); // OK
    pub(); // OK
    priv(); // OK
    privSlot(); // OK
    Q_EMIT privSlot(); // Warning
    Q_EMIT somefunc()->sig(); // OK
    somefunc()->sig(); // Warning
    Q_EMIT memberFunc()->sig(); // OK
    memberFunc()->sig(); // Warning
    emit another->sig(); // OK
    emit s_obj->sig(); // OK
}


MyObject::MyObject()
{
    emit sig(); // Warning
    emit another->sig(); // OK;
    emit memberFunc()->sig(); // OK;
    [this]{ emit sig(); }; // OK
    emit singularSig(); // Warning
    singularSlot(); // OK
}

void MyObject::singularSlot()
{
    singularSig(); // Warning
}

struct NotQObject
{
    QObject *o;
    void test1() {}
    void test()
    {
        test1(); // OK
        emit test1(); // Warning
        emit o->destroyed(); // OK
    }
};

class TestBug373947 : public QObject
{
    int method()
    {
        return otherMethod(); // OK
    }

Q_SIGNALS:
    void someSignal();

public:
    int otherMethod();
};

void wrapping_syntax()
{
    MyObject obj;
    Q_EMIT(obj.sig());
    Q_EMIT((obj.sig()));
    Q_EMIT(( obj.sig() ));
    Q_EMIT(( obj.memberFunc() )); // WARN

    Q_EMIT obj.sig();
    Q_EMIT  obj.sig();
}

