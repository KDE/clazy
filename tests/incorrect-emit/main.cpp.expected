incorrect-emit/main.cpp:39:11: warning: Emit keyword being used with non-signal MyObject::prot [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:40:5: warning: Missing emit keyword on signal call MyObject::sig [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:45:12: warning: Emit keyword being used with non-signal MyObject::privSlot [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:47:5: warning: Missing emit keyword on signal call MyObject::sig [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:49:5: warning: Missing emit keyword on signal call MyObject::sig [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:57:10: warning: Emitting inside constructor probably has no effect [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:61:10: warning: Emitting inside constructor probably has no effect [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:67:5: warning: Missing emit keyword on signal call MyObject::singularSig [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:77:14: warning: Emit keyword being used with non-signal NotQObject::test1 [-Wclazy-incorrect-emit]
incorrect-emit/main.cpp:102:14: warning: Emit keyword being used with non-signal MyObject::memberFunc [-Wclazy-incorrect-emit]
